README.md
<div align="center">
    <h1>Vanilla OS Upgrade Utility</h1>
    <p>This utility is meant to be used in <a href="https://github.com/vanilla-os">Vanilla OS</a> 
    to upgrade the system to the latest version.</p>
</div>


## Build
### Dependencies
- build-essential
- meson
- libadwaita-1-dev
- gettext
- desktop-file-utils

### Build
```bash
meson build
ninja -C build
```

### Install
```bash
sudo ninja -C build install
```

## Run
```bash
vanilla-updater
```


debian/changelog
vanilla-updater (0.0.1-1) lunar; urgency=low

  * First release

 -- Mirko Brombin <send@mirko.pm>  Mon, 20 Dec 2022 15:29:00 +0000

vanilla-updater (0.0.1) kinetic; urgency=low

  * First release

 -- Mirko Brombin <send@mirko.pm>  Mon, 20 Dec 2022 15:29:00 +0000

src/main.py
# main.py
#
# Copyright 2022 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import sys
import gi
import logging

gi.require_version('Gtk', '4.0')
gi.require_version('Adw', '1')

from gi.repository import Gtk, Gio, Adw
from .window import VanillaUpdaterWindow


logging.basicConfig(level=logging.INFO)


class VanillaUpdaterApplication(Adw.Application):
    """The main application singleton class."""

    def __init__(self):
        super().__init__(application_id='vanilla-updater',
                         flags=Gio.ApplicationFlags.FLAGS_NONE)
        self.create_action('quit', self.quit, ['<primary>q'])
        self.create_action('about', self.on_about_action)

    def do_activate(self):
        """Called when the application is activated.

        We raise the application's main window, creating it if
        necessary.
        """
        win = self.props.active_window
        if not win:
            win = VanillaUpdaterWindow(application=self)
        win.present()

    def on_about_action(self, widget, _):
        """Callback for the app.about action."""
        about = Adw.AboutWindow(transient_for=self.props.active_window,
                                application_name='vanilla-updater',
                                application_icon='vanilla-updater',
                                developer_name='Mirko Brombin',
                                version='0.1.0',
                                developers=['Mirko Brombin'],
                                copyright='© 2022 Mirko Brombin')
        about.present()

    def create_action(self, name, callback, shortcuts=None):
        """Add an application action.

        Args:
            name: the name of the action
            callback: the function to be called when the action is
              activated
            shortcuts: an optional list of accelerators
        """
        action = Gio.SimpleAction.new(name, None)
        action.connect("activate", callback)
        self.add_action(action)
        if shortcuts:
            self.set_accels_for_action(f"app.{name}", shortcuts)


def main(version):
    """The application's entry point."""
    app = VanillaUpdaterApplication()
    return app.run(sys.argv)


src/window.py
# window.py
#
# Copyright 2022 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import os
import subprocess
from gi.repository import Adw
from gi.repository import Gtk

from .core.distro import Distro
from .core.updates_repository import UpdatesRepository

from .utils.run_async import RunAsync


@Gtk.Template(resource_path='/org/vanillaos/VanillaUpdater/window.ui')
class VanillaUpdaterWindow(Adw.ApplicationWindow):
    __gtype_name__ = 'VanillaUpdaterWindow'

    stack_main = Gtk.Template.Child()
    button_reboot = Gtk.Template.Child()
    button_upgrade = Gtk.Template.Child()

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.__loader()
        self.button_reboot.connect('clicked', self.__on_reboot_clicked)
        self.button_upgrade.connect('clicked', self.__on_upgrade_clicked)

    def __loader(self):
        def async_fn():
            return Distro(), UpdatesRepository()
        
        def callback(result, *args):
            distro, updates_repo = result
            _version = distro.version

            if "FAKE_VERSION" in os.environ:  # for testing
                _version = os.environ["FAKE_VERSION"]

            update = updates_repo.get_next_update(_version)
            
            if update:
                self.stack_main.set_visible_child_name('found')
            else:
                self.stack_main.set_visible_child_name('updated')

        RunAsync(async_fn, callback)

    def __on_reboot_clicked(self, button):
        subprocess.run(['gnome-session-quit', '--reboot'])

    def __on_upgrade_clicked(self, button):
        proc = subprocess.run(
            ['pkexec', 'vanilla-updater-cli', '--upgrade'], 
            stdout=subprocess.PIPE, 
            env=os.environ.copy()
        )
        if proc.returncode == 0:
            self.stack_main.set_visible_child_name('done')
        else:
            self.stack_main.set_visible_child_name('failed')


src/core/distro.py
# distro.py
#
# Copyright 2022 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import os
import logging


logger = logging.getLogger("VanillaUpdater::Distro")


class Distro:

    def __init__(self):
        self.__id = None
        self.__id_like = None
        self.__name = None
        self.__pretty_name = None
        self.__version = None
        self.__version_id = None
        self.__version_codename = None

        self.__parse_os_release()

    def __parse_os_release(self):
        os_release = open('/etc/os-release', 'r')

        for line in os_release:
            if line.startswith('ID='):
                self.__id = line[3:].strip().replace('"', '')
            elif line.startswith('ID_LIKE='):
                self.__id_like = line[8:].strip().replace('"', '')
            elif line.startswith('NAME='):
                self.__name = line[5:].strip().replace('"', '')
            elif line.startswith('PRETTY_NAME='):
                self.__pretty_name = line[12:].strip().replace('"', '')
            elif line.startswith('VERSION='):
                self.__version = line[8:].strip().replace('"', '')
            elif line.startswith('VERSION_ID='):
                self.__version_id = line[11:].strip().replace('"', '')
            elif line.startswith('VERSION_CODENAME='):
                self.__version_codename = line[17:].strip().replace('"', '')

        os_release.close()
    
    @property
    def dict(self):
        return {
            'id': self.__id,
            'id_like': self.__id_like,
            'name': self.__name,
            'pretty_name': self.__pretty_name,
            'version': self.__version,
            'version_id': self.__version_id,
            'version_codename': self.__version_codename
        }

    @property
    def id(self):
        return self.__id

    @property
    def id_like(self):
        return self.__id_like

    @property
    def name(self):
        return self.__name

    @property
    def pretty_name(self):
        return self.__pretty_name

    @property
    def version(self):
        return self.__version

    @property
    def version_id(self):
        return self.__version_id

    @property
    def version_codename(self):
        return self.__version_codename
        
    @property
    def is_lts(self):
        return "lts" in self.__version.lower().split(" ")


src/core/processor.py
# prepare.py
#
# Copyright 2022 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import os
import json
import requests
import logging
import subprocess

REPO_SOURCES = """deb http://ppa.launchpad.net/vanilla-os/testing/ubuntu {codename} main
# deb-src http://ppa.launchpad.net/vanilla-os/testing/ubuntu {codename} main"""
REPO_LIST_LOCATION = "/tmp/overlayfs-combiner/etc/apt/sources.list.d/vanilla.list"

logger = logging.getLogger("VanillaUpdater::UpgradeTransactionPrepare")


class UpdateProcessor:

    def __init__(self, update: "Update"):
        self.__update = update

    def __check_transactions_lock(self):
        return os.path.exists("/tmp/abroot-transactions.lock")

    def run(self):
        if self.__check_transactions_lock():
            logger.error("Another transaction is running or a reboot is required.")
            return False

        logger.info('Launching upgrade transaction...')

        logger.info('Starting abroot shell...')
        abroot = subprocess.Popen(['abroot', 'shell'], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

        logger.info('Replacing vanilla.list...')
        with open(REPO_LIST_LOCATION, 'w') as f:
            f.write(REPO_SOURCES.format(codename=self.__update.codename))

        logger.info('Updating sources...')
        abroot.stdin.write(b'apt update\n')

        logger.info('Upgrading packages...')
        abroot.stdin.write(b'apt upgrade\n')

        logger.info('Upgrading distro...')
        abroot.stdin.write(b'apt dist-upgrade\n')

        logger.info('Exiting abroot shell...')
        abroot.stdin.write(b'exit\n')

        logger.info('Waiting for abroot shell to exit...')
        abroot.wait()

        if abroot.returncode != 0:
            logger.error('Abroot shell exited with error code {}'.format(abroot.returncode))
            return False

        logger.info('Upgrade transaction terminated.')
        return True


src/utils/run_async.py
# run_async.py
#
# Copyright 2022 brombinmirko <send@mirko.pm>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, in version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import os
import sys
import threading
import traceback
import logging

from gi.repository import GLib

logger = logging.getLogger("VanillaUpdater::Async")


class RunAsync(threading.Thread):
    """
    This class is used to execute a function asynchronously.
    It takes a function, a callback and a list of arguments as input.
    """

    def __init__(self, task_func, callback=None, *args, **kwargs):
        if "DEBUG_MODE" in os.environ:
            import faulthandler
            faulthandler.enable()

        self.source_id = None
        assert threading.current_thread() is threading.main_thread()

        super(RunAsync, self).__init__(
            target=self.__target, args=args, kwargs=kwargs)

        self.task_func = task_func

        self.callback = callback if callback else lambda r, e: None
        self.daemon = kwargs.pop("daemon", True)

        self.start()

    def __target(self, *args, **kwargs):
        result = None
        error = None

        logger.debug(f"Running async job [{self.task_func}].")

        try:
            result = self.task_func(*args, **kwargs)
        except Exception as exception:
            logger.error("Error while running async job: "
                         f"{self.task_func}\nException: {exception}")

            error = exception
            _ex_type, _ex_value, trace = sys.exc_info()
            traceback.print_tb(trace)
            traceback_info = '\n'.join(traceback.format_tb(trace))

        self.source_id = GLib.idle_add(self.callback, result, error)
        return self.source_id

src/core/updates_repository.py
# updates_repository.py
#
# Copyright 2022 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import json
import requests
import logging

UPDATES_URL = 'https://info.vanillaos.org/updates.json'

logger = logging.getLogger("VanillaUpdater::UpdatesRepository")


class Update:
    
    def __init__(self, version, codename):
        self.version = version
        self.codename = codename

    def __str__(self):
        return f'Version: {self.version}, Codename: {self.codename}'

    def __repr__(self):
        return self.__str__()


class UpdatesRepository:

    def __init__(self):
        self.__updates = self.__get_updates()

    def __get_updates(self):
        logger.info('Getting updates...')

        updates = []
        response = requests.get(UPDATES_URL)

        if response.status_code == 200:
            data = json.loads(response.text)
            for version in data:
                updates.append(Update(version, data[version]['codename']))
        
        return updates
    
    def get_next_update(self, version):
        logger.info('Getting next update for version %s', version)

        for update in self.__updates:
            if update.version > version:
                logger.info('Next update: %s', update)
                return update
        
        logger.info('No updates found')
        return None

    @property
    def all_updates(self):
        return self.__updates

    @property
    def latest_update(self):
        return self.__updates[-1]


src/core/prepare.py
# prepare.py
#
# Copyright 2022 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import os
import json
import requests
import logging
import subprocess

REPO_KEY = "https://repo.vanillaos.org/{codename}/KEY.gpg"
REPO_LIST_URL = "https://repo.vanillaos.org/{codename}/vanilla-os.list"
REPO_KEY_LOCATION = "/usr/share/keyrings/vanilla-archive-keyring.gpg"
REPO_LIST_LOCATION = "/etc/apt/sources.list.d/vanilla-os.list"

logger = logging.getLogger("VanillaUpdater::UpgradeTransactionPrepare")


class UpdatePreparer:

    def __init__(self, update: "Update"):
        self.__update = update
        
    def __download_repo_key(self):
        logger.info('Downloading repo key...')
        response = requests.get(REPO_KEY.format(codename=self.__update.codename))
        tmp = '/tmp/vanilla-archive-keyring.gpg'

        if response.status_code == 200:
            if os.path.exists(tmp):
                os.remove(tmp)

            with open(tmp, 'w') as f:
                f.write(response.text)
                return True
                
        logger.error('Failed to download repo key')
        return False

    def __download_repo_list(self):
        logger.info('Downloading repo list...')
        response = requests.get(REPO_LIST_URL.format(codename=self.__update.codename))
        tmp = '/tmp/vanilla-os.list'

        if response.status_code == 200:
            if os.path.exists(tmp):
                os.remove(tmp)

            with open(tmp, 'w') as f:
                f.write(response.text)
                return True
                
        logger.error('Failed to download repo list')
        return False

    def __install_repo_key(self):
        logger.info('Installing repo key...')
        tmp_key = '/tmp/vanilla-archive-keyring.gpg'
        
        proc = subprocess.run(['gpg', '--dearmor', tmp_key])
        if proc.returncode != 0:
            logger.error('Failed to dearmor repo key')
            return False

        proc = subprocess.run(['sudo', 'mv', tmp_key, REPO_KEY_LOCATION])
        if proc.returncode != 0:
            logger.error('Failed to move repo key')
            return False

        return True

    def __install_repo_list(self):
        logger.info('Installing repo list...')
        tmp_list = '/tmp/vanilla-os.list'

        if os.path.exists(REPO_LIST_LOCATION):
            os.rename(REPO_LIST_LOCATION, REPO_LIST_LOCATION + '.bak')

        proc = subprocess.run(['sudo', 'mv', tmp_list, REPO_LIST_LOCATION])
        if proc.returncode != 0:
            logger.error('Failed to move new repo list')
            return False

        return True

    def __update_repo(self):
        logger.info('Updating repo...')
        proc = subprocess.run(['sudo', 'apt', 'update'])
        if proc.returncode != 0:
            logger.error('Failed to update repo')
            return False

        return True

    def __download_updates(self):
        logger.info('Downloading updates...')
        proc = subprocess.run(['pkcon', 'update', '--only-download', '--noninteractive'])
        if proc.returncode != 0:
            logger.error('Failed to download updates')
            return False

        return True

    def run(self):
        logger.info('Preparing upgrade transaction...')

        if not self.__download_repo_key():
            return False

        if not self.__download_repo_list():
            return False

        if not self.__install_repo_key():
            return False

        if not self.__install_repo_list():
            return False

        if not self.__update_repo():
            return False

        if not self.__download_updates():
            return False

        logger.info('Upgrade transaction prepared')
        return True
